server/grpc_service.go

```diff
@@ -176,6 +161,137 @@ func (s *Server) Tso(stream pdpb.PD_TsoServer) error {
 	}
 }
 
+type tsoRequest struct {
+	forwardedHost string
+	request       *pdpb.TsoRequest
+	stream        pdpb.PD_TsoServer
+}
+
+func (s *Server) dispatchTSORequest(ctx context.Context, request *tsoRequest, forwardedHost string, doneCh <-chan struct{}, errCh chan<- error) {
+	tsoRequestChInterface, loaded := s.tsoDispatcher.LoadOrStore(forwardedHost, make(chan *tsoRequest, maxMergeTSORequests))
+	tsoRequestCh := tsoRequestChInterface.(chan *tsoRequest)
+	if !loaded {
+		go s.handleDispatcher(ctx, forwardedHost, tsoRequestCh, doneCh, errCh)
+	}
+	tsoRequestCh <- request
+}
+
+func (s *Server) handleDispatcher(ctx context.Context, forwardedHost string, tsoRequestCh <-chan *tsoRequest, doneCh <-chan struct{}, errCh chan<- error) {
+	dispatcherCtx, ctxCancel := context.WithCancel(ctx)
+	defer ctxCancel()
+	defer s.tsoDispatcher.Delete(forwardedHost)
+
+	var (
+		forwardStream pdpb.PD_TsoClient
+		cancel        context.CancelFunc
+	)
+	client, err := s.getDelegateClient(ctx, forwardedHost)
+	if err != nil {
+		goto errHandling
```

Comment:
Why not send err to errCh and return?

Reply:
Because the error handling logic is the same as the `createTsoForwardStream` right below it, and I don't want to write the same code twice.